python/ray/gateway.py

```diff
@@ -0,0 +1,110 @@
+import click
+from flask import Flask, request, send_file
+import io
+import pickle
+import pyarrow
+from pyarrow import plasma as plasma
+import sys
+
+app = Flask(__name__)
+ctx = pyarrow.default_serialization_context()
```

Comment:
There are going to be some issues with this potentially.

Currently, we register custom serializers/deserializers in the serialization context for non-primitive objects. This is done under the hood, e.g., 

https://github.com/ray-project/ray/blob/8e687cbc9838fce0b1d8190c7851b7d573562487/python/ray/worker.py#L305-L339

Basically, a worker tries to serialize an object that it doesn't know how to serialize. Pyarrow raises an exception. We attempt to define a custom serializer on the fly. We broadcast that custom serializer to all the other workers/drivers so they know how to serialize that object also.

When we transfer objects from the gateway to the client, we probably want to use `ray.cloudpickle`. However, the gateway may need to also subscribe to the relevant Redis channel in order to get the custom functions needed to deserialize custom objects.

It'd be ok to not support this in a first pass, but we should make sure that we understand the issue and that we're catching the error as anticipated.

Reply:
Actually, maybe we don't want to do any deserialization in the gateway and we only want to pass along the bytes..

Reply:
Yeah, ideally, we don't want to do any serialization or deserialization. We're just shuffling bytes to and from the other ```ray``` processes on the head node. There are some issues around ```numpy``` arrays stored in shared memory, but that's only tangentially related.